### PR TITLE
Payment config admin

### PR DESCRIPTION
**Pull request by arjun289**
Payment Method Configuration in Admin Application.
## Motivation and Context
Admin application should have a way to configure different payment methods such as credit card,
hosted payments, and store credits. It allows the admin to configure payment methods which should be visible to the user to choose from.

The change at present handles the creation of payment methods of types which are fixed in the system(cod, credit card, hosted payment, store credit). Please note this doesn't allow the creation of new payment types in the system as that would require handling subtype and supertype relationship and other details. 

The changes in this PR only allows configuring the existing payment types:
- Payment provider for that particular method 
- Configuring credentials if payment provider is a gateway.
- Whether the payment method is active or not.
- Configuring if the live mode or test url would be used for trying the transaction.

## Describe your changes

- PaymentMethod creation
  Allows creating new payment methods that would be
  available for selection by the customer.

  The payment types for example cash on delivery, credit card,
  store credits are constant at present in the system. New payment methods no way means these types can also be created.
- Adds functionality to perform the usual update and delete operations.
- Adds new fields to the payment method table.

- Migration
  Migration to alter the payment method table adding new field
  and dropping constraint for unique code.
  20180801110537_alter_payment_method_fields.exs
- 

## Any further comments?
The PR also fixes a **bug**.
There were some bugs around names shown on the front end and a bug related to no association
constraint whenever a permission was deleted when it is associated with a role.
Multiple select roles for permission was very basic.
- Fixed the bugs around names and also fixed the no assoc constraint bug. 
- Modified multi-select to use select2js.

**Snitch Payments**
Earlier SnitchPayments was part of the umbrella app. But, due to some unknown errors on Travis, it was moved outside as a separate library. It is now added as a dependency in admin application.

------

## Checklist

- [x] I have read [CONTRIBUTING.md][contributing].
- [x] My code follows the [style guidelines][contributing] of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My code does not generate any (new) [`credo`][credo] and compile-time warnings.
- [x] I have updated the documentation wherever necessary.
- [ ] I have added tests to cover my changes.

[contributing]: https://github.com/aviabird/snitch/CONTRIBUTING.md
[credo]: https://github.com/rrrene/credo

------
Dear Gatekeeper,

Please make sure that the commits that will be merged or rebased into the base
branch are [formatted according to our standards][commit-style].

> The only additional requirement is that the PR number must appear in the
> commit title in brackets: `(#XXX)`

[commit-style]: https://github.com/aviabird/snitch/blob/develop/CONTRIBUTING.md#styling

### CODE REVIEW

File: apps/admin_app/assets/js/app.js [hunk 1]
```diff
@@ -22,8 +22,10 @@ import "phoenix_html"
 
 import "bootstrap";
 import select2Selector from './form-helpers/select2-selector';
+import {getPaymentMethod} from './payment_method'
```
Comment (pkrawat1) [inline apps/admin_app/assets/js/app.js]: spacing after/before {}

File: apps/admin_app/lib/admin_app_web/controllers/permission_controller.ex [hunk 2]
```diff
@@ -83,4 +88,11 @@ defmodule AdminAppWeb.PermissionController do
       description: permission["description"]
     }
   end
+
+  def stringify_error(error_map) do
+    Enum.reduce(error_map, "", fn {key, value}, acc ->
+      value = Enum.join(value, ",")
+      acc <> " " <> to_string(key) <> ": " <> value
```
Comment (pkrawat1) [inline apps/admin_app/lib/admin_app_web/controllers/permission_controller.ex]: `"#{acc} #{key}: #{value}"`

File: apps/admin_app/lib/admin_app_web/templates/payment_method/edit.html.eex [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  <hr/>
+  <%= form_for @changeset, payment_method_path(@conn, :update, @payment_method), [as: :payment_method], 
+    fn f -> %>
+    <%= select_input(f, :provider, providers(),
+        "provider", value: @payment_method.provider, id: "payment-provider",
+        prompt: "Choose provider") %>
+    <%= input(f, :name)%>
+    <%= input(f, :description) %>
+    <div id="preferences">
+      <%= render "preferences_form.html", credentials: @payment_method.preferences %>
```
Comment (pkrawat1) [inline apps/admin_app/lib/admin_app_web/templates/payment_method/edit.html.eex]: prefer `preferences_inputs.html`

File: apps/admin_app/lib/admin_app_web/templates/role/_form.html.eex [hunk 4]
```diff
@@ -1,8 +1,11 @@
 <%= form_for @changeset, @action, [as: :role], fn f -> %>
   <%= input f, :name %>
   <%= input f, :description %>
+  
+  <%= label f, :permissions, "Permissions" %>
   <%= multiple_select(f, :permissions, formatted_list(:permissions),
-        class: "form-control custom-select")%>
+        prompt: "Add Permission", class: "full-width w-100", "data-init-plugin": "select2")%>
```
Comment (pkrawat1) [inline apps/admin_app/lib/admin_app_web/templates/role/_form.html.eex]: Multiselect input helper missing

File: apps/admin_app/lib/admin_app_web/views/payment_method_view.ex [hunk 5]
```diff
@@ -0,0 +1,18 @@
+defmodule AdminAppWeb.PaymentMethodView do
+  use AdminAppWeb, :view
+  alias SnitchPayments
+
+  def capitalize(data) when is_atom(data) do
+    data
+    |> Atom.to_string()
+    |> String.capitalize()
+  end
+
```
Comment (pkrawat1) [inline apps/admin_app/lib/admin_app_web/views/payment_method_view.ex]: no spacing between same methods

File: apps/snitch_core/config/test.exs [hunk 6]
```diff
@@ -7,6 +7,8 @@ config :snitch_core, Snitch.Repo,
   password: "postgres",
   database: "snitch_test",
   hostname: "localhost",
+  ownership_timeout: 60_000,
+  pool_timeout: 60_000,
```
Comment (pkrawat1) [inline apps/snitch_core/config/test.exs]: remove these

File: apps/snitch_core/lib/core/data/schema/payment/payment_method.ex [hunk 7]
```diff
@@ -14,12 +14,18 @@ defmodule Snitch.Data.Schema.PaymentMethod do
     field(:code, :string, size: 3)
     field(:active?, :boolean, default: true)
 
+    field(:live_mode?, :boolean, default: false)
```
Comment (pkrawat1) [inline apps/snitch_core/lib/core/data/schema/payment/payment_method.ex]: add field docs

File: apps/admin_app/lib/admin_app_web/templates/role/_form.html.eex (hunk 4 above)
Comment (arjun289) [reply apps/admin_app/lib/admin_app_web/templates/role/_form.html.eex]: I will add an issue for this, I will pick this up later on.

**Pull request merged**